frame/contracts/src/exec.rs

```diff
@@ -1334,16 +1334,34 @@ where
 		&mut self.top_frame_mut().nested_gas
 	}
 
-	fn append_debug_buffer(&mut self, msg: &str) -> Result<bool, DispatchError> {
+	fn append_debug_buffer(&mut self, msg: &str) -> bool {
 		if let Some(buffer) = &mut self.debug_message {
-			if !msg.is_empty() {
-				buffer
-					.try_extend(&mut msg.bytes())
-					.map_err(|_| Error::<T>::DebugBufferExhausted)?;
-			}
-			Ok(true)
+			let mut msg = msg.bytes();
+			let num_drain = {
+				let capacity = DebugBufferVec::<T>::bound().checked_sub(buffer.len()).expect(
+					"
+					`buffer` is of type `DebugBufferVec`,
+					`DebugBufferVec` is a `BoundedVec`,
+					`BoundedVec::bound()` <= `BoundedVec::len()`;
+					qed
+				",
+				);
+				msg.len().saturating_sub(capacity).min(buffer.len())
+			};
+			buffer.drain(0..num_drain);
+			buffer
+				.try_extend(&mut msg)
+				.map_err(|_| {
+					log::debug!(
```

Comment:
So if `msg.len() > BoundedVec::bound()`,

then we drain all the debug buffer, and msg won't get into it, right?

What about putting this log message to the debug buffer, instead?

Reply:
I didn't include it on purpose. I don't think we want to allow users to post arbitrarily long messages to the logging system. For the same reason that we limit the debug buffer. Maybe it can trigger some memory exhaustion. Given, the size of a single message is limited by the linear memory of a contract which makes it less dangerous. However, I think this corner case is not important enough to create this potential foot gun: The assumptions could change in the future.

Reply:
Yeah I agree with your point. But what I meant here is why not just add your _error_ message to the buffer instead of cleaning the buffer entirely (killing all possibly valuable messages the buffer had had to moment). Suggested this little fix [here](https://github.com/paritytech/substrate/pull/12993/commits/0f04a5c429d88be3d873120e17d78e9d79039b9a#diff-98cc07c06a4567de1a156035c46689e994f2436a5def2d998c91146810ba1cc4L1337).